File: forge/plugin/opencode/src/governor.ts
import { createHash } from "node:crypto";
import { existsSync, realpathSync } from "node:fs";
import { dirname, isAbsolute, relative, resolve } from "node:path";

export const GovernorMode = {
  OFF: "off",
  REPORT: "report",
  ACTIVE: "active",
} as const;

export type GovernorMode = (typeof GovernorMode)[keyof typeof GovernorMode];

const VALID_MODES = new Set<string>(Object.values(GovernorMode));
const DUPLICATE_TOOLS = new Set(["read", "grep", "glob"]);
const PATH_KEYS = new Set([
  "path",
  "filepath",
  "file",
  "filename",
  "cwd",
  "directory",
  "destination",
  "target",
]);

export interface GovernorCapabilities {
  can_block_before: boolean;
  can_replace_output: boolean;
  can_request_confirmation: boolean;
}

export interface GovernorDecision {
  schema_version: number;
  ok: boolean;
  decision: "allow" | "warn" | "escalate" | "block";
  reason: string;
  capability_limited: boolean;
}

function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([key, child]) => [key, canonicalize(child)]),
    );
  }
  return value;
}

export function fingerprint(toolName: string, args: Record<string, unknown>): string {
  const wire = JSON.stringify({
    arguments: canonicalize(args),
    tool: toolName.trim().toLowerCase(),
  });
  return createHash("sha256").update(wire).digest("hex");
}

function resolvedWithExistingAncestors(candidate: string): string {
  let probe = candidate;
  const suffix: string[] = [];
  while (!existsSync(probe)) {
    const parent = dirname(probe);
    if (parent === probe) return resolve(candidate);
    suffix.unshift(relative(parent, probe));
    probe = parent;
  }
  return resolve(realpathSync(probe), ...suffix);
}

function escapesRoot(candidate: string, repoRoot: string): boolean {
  const rel = relative(repoRoot, candidate);
  return rel === ".." || rel.startsWith(`..${process.platform === "win32" ? "\\" : "/"}`) || isAbsolute(rel);
}

export function unsafePaths(value: unknown, repoRoot: string): string[] {
  const found: string[] = [];

  function visit(item: unknown, key: string): void {
    if (item === null || item === undefined) return;
    if (Array.isArray(item)) {
      for (const child of item) visit(child, key);
      return;
    }
    if (typeof item === "object") {
      for (const [childKey, child] of Object.entries(item as Record<string, unknown>)) {
        visit(child, childKey.toLowerCase());
      }
      return;
    }
    if (typeof item !== "string" || !PATH_KEYS.has(key)) return;
    if (item.includes("\0")) { found.push(item); return; }
    try {
      const joined = isAbsolute(item) ? resolve(item) : resolve(repoRoot, item);
      if (escapesRoot(resolvedWithExistingAncestors(joined), repoRoot)) found.push(item);
    } catch {
      found.push(item);
    }
  }

  visit(value, "");
  return found;
}

export function dangerousCommandReason(command: string): string | null {
  const normalized = command.toLowerCase().replace(/\s+/g, " ").trim();
  if (!normalized) return null;

  if (/(^|[;&|]\s*|\s)(?:command\s+)?rm\s+(?:(?:-[^\s]*[rf][^\s]*|--recursive|--force)\s+){2,}/.test(normalized)) {
    return "recursive forced removal requires approval";
  }
  if (/(^|[;&|]\s*|\s)(?:command\s+)?rm\s+-[^\s]*r[^\s]*f[^\s]*(?:\s|$)/.test(normalized)
      || /(^|[;&|]\s*|\s)(?:command\s+)?rm\s+-[^\s]*f[^\s]*r[^\s]*(?:\s|$)/.test(normalized)) {
    return "recursive forced removal requires approval";
  }
  if (/\bgit(?:\s+-c\s+\S+|\s+-c\S+|\s+-C\s+\S+)*\s+(?:reset\s+--hard|clean\s+-[^\s]*f|push\b[^;&|]*--force(?:-with-lease)?)/i.test(command)) {
    return "destructive Git command requires approval";
  }
  if (/(^|[;&|]\s*|\s)(sudo|mkfs(?:\.\w+)?|shutdown|reboot|poweroff)(?:\s|$)/.test(normalized)) {
    return "privileged or system command requires approval";
  }
  if (/(^|[;&|]\s*|\s)(dd|chmod|chown)\s+/.test(normalized)) {
    return "high-impact filesystem command requires approval";
  }
  return null;
}

export class ContextGovernor {
  private readonly mode: GovernorMode;
  private readonly repoRoot: string;
  private readonly capabilities: GovernorCapabilities;
  private readonly recentBySession = new Map<string, Array<[number, string]>>();
  private readonly duplicateCount: number;
  private readonly duplicateSeconds: number;

  constructor(
    mode: GovernorMode | string,
    repoRoot: string,
    capabilities: Partial<GovernorCapabilities> = {},
    duplicateCount = 16,
    duplicateSeconds = 60,
  ) {
    if (!VALID_MODES.has(String(mode))) throw new Error(`invalid governor mode: ${mode}`);
    this.mode = String(mode) as GovernorMode;
    this.repoRoot = resolvedWithExistingAncestors(resolve(repoRoot));
    this.capabilities = {
      can_block_before: false,
      can_replace_output: false,
      can_request_confirmation: false,
      ...capabilities,
    };
    this.duplicateCount = duplicateCount;
    this.duplicateSeconds = duplicateSeconds;
  }

  before(sessionId: string, toolName: string, args: Record<string, unknown>): GovernorDecision {
    if (!sessionId || !toolName || typeof args !== "object" || args === null) {
      return this.decision("block", "invalid governor input", "can_block_before");
    }
    if (this.mode === GovernorMode.OFF) return this.decision("allow", "governor is off");

    const normalizedTool = toolName.trim().toLowerCase();
    if (DUPLICATE_TOOLS.has(normalizedTool)) {
      const duplicate = this.trackDuplicate(sessionId, fingerprint(normalizedTool, args));
      if (duplicate) {
        const action = this.mode === GovernorMode.ACTIVE ? "block" : "warn";
        return this.decision(
          action,
          "duplicate call blocked. Retrieve the previous result with forge_expand_output " +
            "using the handle from the earlier output, or vary your arguments.",
          "can_block_before",
        );
      }
    }

    const command = String(args.command ?? args.cmd ?? "");
    const dangerous = dangerousCommandReason(command);
    if (dangerous) {
      const action = this.mode === GovernorMode.ACTIVE ? "escalate" : "warn";
      return this.decision(action, dangerous, "can_request_confirmation");
    }

    const unsafe = unsafePaths(args, this.repoRoot);
    if (unsafe.length > 0) {
      const action = this.mode === GovernorMode.ACTIVE ? "escalate" : "warn";
      return this.decision(
        action,
        `path is outside the controlled repository: ${unsafe[0]}`,
        "can_request_confirmation",
      );
    }

    return this.decision("allow", "no policy concern detected");
  }

  clearSession(sessionId: string): void {
    this.recentBySession.delete(sessionId);
  }

  private trackDuplicate(sessionId: string, key: string): boolean {
    const now = Date.now() / 1000;
    const recent = this.recentBySession.get(sessionId) ?? [];
    while (
      recent.length > 0
      && (recent.length >= this.duplicateCount || now - recent[0][0] > this.duplicateSeconds)
    ) {
      recent.shift();
    }
    const duplicate = recent.some(([, existing]) => existing === key);
    recent.push([now, key]);
    this.recentBySession.set(sessionId, recent);
    return duplicate;
  }

  private decision(
    requested: GovernorDecision["decision"],
    reason: string,
    needs?: keyof GovernorCapabilities,
  ): GovernorDecision {
    let decision = requested;
    let capabilityLimited = false;
    if (this.mode === GovernorMode.REPORT && decision !== "allow" && decision !== "warn") {
      decision = "warn";
    }
    if (this.mode === GovernorMode.ACTIVE && needs && !this.capabilities[needs]) {
      capabilityLimited = true;
      decision = "warn";
      reason += `; adapter lacks ${needs}`;
    }
    return {
      schema_version: 1,
      ok: true,
      decision,
      reason,
      capability_limited: capabilityLimited,
    };
  }
}

File: forge/context/governor.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from enum import Enum
import hashlib
import json
from pathlib import Path
import re
import time
from typing import Any, Callable

from .formatter import estimate_tokens
from .result_store import ToolResultStore


class GovernorMode(str, Enum):
    OFF = "off"
    REPORT = "report"
    ACTIVE = "active"


@dataclass(frozen=True, slots=True)
class GovernorCapabilities:
    can_block_before: bool = False
    can_replace_output: bool = False
    can_request_confirmation: bool = False


DANGEROUS = (
    re.compile(r"(^|\s)rm\s+(-[^\s]*r[^\s]*f|-[^\s]*f[^\s]*r)\b"),
    re.compile(r"\bgit\s+(reset\s+--hard|clean\s+-[^\s]*f|push\s+.*--force)\b"),
    re.compile(r"\b(?:sudo|mkfs|shutdown|reboot)\b"),
)


def fingerprint(tool_name: str, arguments: dict[str, Any]) -> str:
    wire = json.dumps({"tool": tool_name.strip().lower(), "arguments": arguments},
                      sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return hashlib.sha256(wire.encode()).hexdigest()


class ContextGovernor:
    def __init__(self, mode: GovernorMode | str, repo_root: Path, result_store: ToolResultStore,
                 capabilities: GovernorCapabilities | None = None,
                 clock: Callable[[], float] = time.time, duplicate_count: int = 16,
                 duplicate_seconds: float = 60, large_output_tokens: int = 2000) -> None:
        self.mode = GovernorMode(mode)
        self.repo_root = repo_root.resolve()
        self.result_store = result_store
        self.capabilities = capabilities or GovernorCapabilities()
        self.clock = clock
        self.duplicate_count = duplicate_count
        self.duplicate_seconds = duplicate_seconds
        self.large_output_tokens = large_output_tokens
        self._recent: deque[tuple[float, str]] = deque()

    def before(self, task_id: str, tool_name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        if not task_id or not tool_name or not isinstance(arguments, dict):
            return self._decision("block", "invalid governor input", needs="can_block_before")
        if self.mode is GovernorMode.OFF:
            return self._decision("allow", "governor is off")
        now = self.clock()
        key = fingerprint(tool_name, arguments)
        while self._recent and (len(self._recent) >= self.duplicate_count or now - self._recent[0][0] > self.duplicate_seconds):
            self._recent.popleft()
        duplicate = any(existing == key for _, existing in self._recent)
        self._recent.append((now, key))
        if duplicate:
            action = "block" if self.mode is GovernorMode.ACTIVE else "warn"
            return self._decision(action, "exact duplicate tool call", needs="can_block_before")
        command = str(arguments.get("command", arguments.get("cmd", "")))
        if any(pattern.search(command) for pattern in DANGEROUS):
            return self._decision("escalate" if self.mode is GovernorMode.ACTIVE else "warn",
                                  "dangerous command requires user confirmation",
                                  needs="can_request_confirmation")
        unsafe = self._unsafe_paths(arguments)
        if unsafe:
            return self._decision("escalate" if self.mode is GovernorMode.ACTIVE else "warn",
                                  "path is outside the controlled repository: " + unsafe[0],
                                  needs="can_request_confirmation")
        return self._decision("allow", "no policy concern detected")

    def after(self, task_id: str, tool_name: str, output: str) -> dict[str, Any]:
        if not isinstance(output, str):
            return self._decision("warn", "tool output was not text")
        if self.mode is GovernorMode.OFF or estimate_tokens(output) <= self.large_output_tokens:
            return self._decision("allow", "output is within limit")
        if self.mode is GovernorMode.REPORT:
            return self._decision("warn", "large tool output detected")
        if not self.capabilities.can_replace_output:
            return self._decision("warn", "large output cannot be replaced by this adapter",
                                  capability_limited=True)
        handle = self.result_store.store(task_id, output)
        replacement = f"Large output stored as {handle}. Use forge_expand_tool_result with this task id."
        return self._decision("replace", "large output stored for bounded expansion",
                              replacement_output=replacement, handle=handle)

    def _unsafe_paths(self, value: Any) -> list[str]:
        found: list[str] = []
        path_keys = {"path", "file", "filename", "filepath", "cwd", "directory", "destination", "target"}

        def visit(item: Any, key: str = "") -> None:
            if isinstance(item, dict):
                for child_key, child in item.items():
                    visit(child, str(child_key).lower())
            elif isinstance(item, list):
                for child in item:
                    visit(child, key)
            elif isinstance(item, str) and key in path_keys:
                candidate = Path(item)
                target = candidate if candidate.is_absolute() else self.repo_root / candidate
                try:
                    resolved = target.resolve(strict=False)
                    if not resolved.is_relative_to(self.repo_root):
                        found.append(item)
                    elif target.exists() and target.resolve(strict=True).is_relative_to(self.repo_root) is False:
                        found.append(item)
                except (OSError, ValueError):
                    found.append(item)

        visit(value)
        return found

    def _decision(self, decision: str, reason: str, needs: str | None = None,
                  capability_limited: bool = False, **extra: Any) -> dict[str, Any]:
        if self.mode is GovernorMode.REPORT and decision not in {"allow", "warn"}:
            decision = "warn"
        if self.mode is GovernorMode.ACTIVE and needs and not getattr(self.capabilities, needs):
            capability_limited = True
            decision = "warn"
            reason += f"; adapter lacks {needs}"
        return {"schema_version": 1, "ok": True, "decision": decision, "reason": reason,
                "replacement_output": extra.pop("replacement_output", None),
                "capability_limited": capability_limited, **extra}

